sources/sqlite3_cql_extension/cqlsqlite3extension.py

```diff
@@ -0,0 +1,355 @@
+        print(f"""
+  rc = sqlite3_create_function(db, "{cmd_args['namespace']}{proc['name']}", {len([arg for arg in proc['args'] if arg.get('binding', 'in') in ['in', 'inout']])}, SQLITE_UTF8, NULL, call_{cmd_args['namespace']}{proc['name']}, NULL, NULL);
+  if (rc != SQLITE_OK) return rc;
+""")
+
+    print("""
+  return rc;
+}
+
+""")
+
+def emit_all_procs(data):
+    for proc in (
+        data["queries"] +
+        data["deletes"] +
+        data["inserts"] +
+        data["generalInserts"] +
+        data["updates"] +
+        data["general"]
+    ):
+        proc['usesDatabase'] = proc["usesDatabase"] if "usesDatabase" in proc else True
+        proc['projection'] = "projection" in proc
+        proc['canonicalName'] = f"call_{cmd_args['namespace']}{proc['name']}"
+
+        attributes = {}
+        for attr in proc.get("attributes", []):
+            attributes[attr['name']] = attr['value']
+
+        # no codegen for private methods
+        if "cql:private" in attributes: return
+
+        emit_proc_c_func_body(proc, attributes)
+
+# This emits the main body of the C Interop function, this includes
+# * the Interop entry point for the procedure
+# * the call to the procedure
+# * the marshalling of the results
+# * the return of the results
+# * the cleanup of the results
+def emit_proc_c_func_body(proc, attributes):
+    in_arguments = [arg for arg in proc['args'] if arg.get('binding', 'in') == ('in')]
+    inout_arguments = [arg for arg in proc['args'] if arg.get('binding', 'in') == ('inout')]
+    out_arguments = [arg for arg in proc['args'] if arg.get('binding', 'in') == ('out')]
+
+    innie_arguments = in_arguments + inout_arguments
+    outtie_arguments = inout_arguments + out_arguments
+
+
+    print(f"void {proc['canonicalName']}(sqlite3_context *_Nonnull context, int32_t argc, sqlite3_value *_Nonnull *_Nonnull argv)", end="")
+    print("{")
+
+    print(f"  // 1. Ensure Sqlite function argument count matches count of the procedure in and inout arguments")
+    print(f"  if (argc != {len(innie_arguments)}) goto invalid_arguments;")
+    print(f"")
+
+
+    print("  // 2. Ensure sqlite3 value type is compatible with cql type")
+    for index, arg in enumerate(innie_arguments):
+        print(
+            f"  if (!is_sqlite3_type_compatible_with_cql_core_type("
+                f"sqlite3_value_type(argv[{index}]), "
+                f"{row_types[arg['type']]}, "
+                f"{'cql_false' if arg['isNotNull'] else 'cql_true'}"
+            f")) goto invalid_arguments; // {arg['name']}"
+        )
+    print("")
+
+
+    print("  // 3. Marshalled argument initialization")
+    for index, arg in enumerate(innie_arguments):
+        if arg['type'] in ("text", "blob", "object"):
+            print(f"  {cql_nullable_types[arg['type']].ljust(20)} {arg['name'].ljust(25)} = {sqlite3_value_getter_nullable[arg['type']]}(argv[{index}]);")
```

Comment:
I mean `ljust`? That seems like overkill in generated code but if it helps you to debug who am I to say no. It's not like it will cause merge conflicts...

Reply:
I don't like having having it in the python script but I find it helpful when reading the output.
Since it's generated, you never have to manually maintain the spacing.

<img width="500" alt="image" src="https://github.com/user-attachments/assets/866c2e85-2e46-4650-975d-bcce809a80f5">

<img width="500" alt="image" src="https://github.com/user-attachments/assets/b8236d65-35ee-4889-8957-cd659ea5af13">